File: testing/unit-test-patterns/javascript/src/test-data-builders/builders/cloneTestDataBuilder.ts
import { IClone } from "../../common/interfaces/Clone.interface";
import { Gender } from "../../common/models/gender";

export class CloneTestDataBuilder {  
  private firstName: string
  private lastName: string
  private gender: Gender
  private numberOfEyes: number
  private hairColor: string

  constructor() {
    this.firstName = ''
    this.lastName = ''
    this.gender = Gender.male
    this.numberOfEyes = 0
    this.hairColor = ''
  }

  withGenericMaleBase() {
    this.firstName = 'John'
    this.lastName = 'Smith'
    this.gender = Gender.male
    this.numberOfEyes = 2
    this.hairColor = 'Black'

    return this
  }

  withFirstName(firstName: string) {
    this.firstName = firstName
    return this
  }

  withLastName(lastName: string) {
    this.lastName = lastName
    return this
  }

  withNumberOfEyes(numberOfEyes: number) {
    this.numberOfEyes = numberOfEyes
    return this
  }

  build(): IClone {
    return {
      firstName: this.firstName,
      lastName: this.lastName,
      gender: this.gender,
      numberOfEyes: this.numberOfEyes,
      hairColor: this.hairColor,
    }
  }
}

File: testing/unit-test-patterns/c#/UnitTestPatterns/TestDataBuilder/Builders/CloneTestDataBuilder.cs
using NSubstitute;
using UnitTestPatterns.Common.Interfaces;
using UnitTestPatterns.Common.Models;

namespace UnitTestPatterns.TestDataBuilder.Builders
{
    public class CloneTestDataBuilder
    {
        private readonly IClone _clone;

        public CloneTestDataBuilder()
        {
            _clone = Substitute.For<IClone>();
        }

        public IClone Build()
        {
            return _clone;
        }

        public CloneTestDataBuilder WithFirstName(string firstName)
        {
            _clone.FirstName.Returns(firstName);
            return this;
        }

        public CloneTestDataBuilder WithLastName(string lastName)
        {
            _clone.LastName.Returns(lastName);
            return this;
        }

        private CloneTestDataBuilder WithGender(Gender gender)
        {
            _clone.Gender.Returns(gender);
            return this;
        }

        private CloneTestDataBuilder WithHairColor(string hairColor)
        {
            _clone.HairColor.Returns(hairColor);
            return this;
        }

        public CloneTestDataBuilder WithStandardNumberOfEyes()
        {
            _clone.NumberOfEyes.Returns(2);
            return this;
        }

        public CloneTestDataBuilder IsTriclops()
        {
            _clone.NumberOfEyes.Returns(3);
            return this;
        }

        public CloneTestDataBuilder WithGenericMaleBase()
        {
            return WithFirstName("John")
                .WithLastName("Smith")
                .WithGender(Gender.Male)
                .WithHairColor("Black")
                .WithStandardNumberOfEyes();
        }

        public IClone BuildGenericMale()
        {
            return WithGenericMaleBase()
                .Build();
        }

        public IClone BuildGenericFemale()
        {
            return WithFirstName("Jane")
                .WithLastName("Doe")
                .WithGender(Gender.Female)
                .WithHairColor("Blonde")
                .WithStandardNumberOfEyes()
                .Build();
        }
    }
}